framework/src/tinyalsa/tinyalsa.c

```diff
@@ -673,6 +683,14 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 					}
 				}
 
+			} else if (msg.msgId == AUDIO_MSG_XRUN) {
```

Comment:
Shouldn't tinyalsa do something more, like idenfitying semi filled buffer, and remove if before continuing?

Reply:
According to tinyalsa code, when underrun happens -->
- Restart pcm
- Try to redo the read / write operation

This is what I am doing here. pcm_stop -> start will restart the pcm. This will clear all pending buffers and enqueue new buffers. Also, I am using continue keyword. So, in next iteration of the loop, we will retry to perform read operation.

Reply:
But what about the half filled buffer (in case)? How is it handled in tinyalsa? (just ignored?)